components/de_amp/browser/de_amp_url_loader.h

```diff
@@ -0,0 +1,137 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_DE_AMP_BROWSER_DE_AMP_URL_LOADER_H_
+#define BRAVE_COMPONENTS_DE_AMP_BROWSER_DE_AMP_URL_LOADER_H_
+
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "base/callback.h"
+#include "base/memory/ref_counted.h"
+#include "base/memory/weak_ptr.h"
+#include "brave/components/de_amp/browser/de_amp_service.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/pending_receiver.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "mojo/public/cpp/bindings/receiver.h"
+#include "mojo/public/cpp/bindings/remote.h"
+#include "mojo/public/cpp/system/data_pipe.h"
+#include "mojo/public/cpp/system/simple_watcher.h"
+#include "services/network/public/mojom/early_hints.mojom-forward.h"
+#include "services/network/public/mojom/url_loader.mojom.h"
+#include "services/network/public/mojom/url_response_head.mojom-forward.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace de_amp {
+
+class DeAmpThrottle;
+
+class DeAmpURLLoader : public network::mojom::URLLoaderClient,
```

Comment:
ОПРЕДЕЛИТЬ ЯЗЫК

АНГЛИЙСКИЙ

РУССКИЙ

НЕМЕЦКИЙ

РУССКИЙ

АНГЛИЙСКИЙ

УКРАИНСКИЙ
Исходный текст
Are DeAmpURLLoader/DeAmpThrottle copy/paste from SpeedReaderURLLoader/SpeedReaderThrottle?
If we need to check request bodies then better to make one general throttle/loader for such purposes, with the ability to subscribe/control it via some interface.
This:
1) reduce code and test duplication;
2) avoid excessive performance impact, because throttles work one after another (but this check, called after speedreader, will be faster than usual).

Reply:
It would be good to refactor the existing throttle, but I think we should do that once this is merged in and shipping - at that point we'll have a better idea of what needs this particular throttle has.

Reply:
In this PR we have a lot of similar code. Also it isn't a simple task to cover this code by tests, because of a lot of states.
 If we merge this it will be hard to maintain and we will have to refactor the tests too.

All such throttles have a clear purpose: to spy on a loading document body.